File: easy/1056_confusing_number/solution.cpp
using namespace std;

#include <unordered_map>

class Solution {
public:
  bool confusingNumber(int n) {
    unordered_map<int, int> confusingMap = {
        {0, 0}, {1, 1}, {6, 9}, {8, 8}, {9, 6}};

    const int nCopy = n;
    int nRotated = 0;

    int digit;

    while (n) {
      digit = n % 10;

      n /= 10;
      nRotated *= 10;

      if (confusingMap.find(digit) != confusingMap.end())
        digit = confusingMap[digit];
      else
        return false;

      nRotated += digit;
    }

    return nCopy != nRotated;
  }
};

File: easy/1056_confusing_number/solution.py
class Solution:
    def confusingNumber(self, n: int) -> bool:
        confusing_map = {0: 0, 1: 1, 6: 9, 8: 8, 9: 6}

        n_copy = n
        rotated_n = 0

        while n:
            digit = n_copy % 10

            n_copy //= 10
            rotated_n *= 10

            if digit in confusing_map:
                digit = confusing_map[digit]
            else:
                return False

            rotated_n += digit

        return n != rotated_n
